File: src/Systems/Objective-C/memento.m
#import "memento.h"
#import <stdio.h>

@interface MementoSnapshot : NSObject
@property(nonatomic, copy, readonly) NSString *state;
- (instancetype)initWithState:(NSString *)state;
@end

@implementation MementoSnapshot
- (instancetype)initWithState:(NSString *)state {
    if ((self = [super init])) {
        _state = [state copy];
    }
    return self;
}
@end

@interface MementoDocument : NSObject
@property(nonatomic, copy) NSString *state;
- (instancetype)initWithState:(NSString *)state;
- (MementoSnapshot *)save;
- (void)restore:(MementoSnapshot *)snapshot;
@end

@implementation MementoDocument
- (instancetype)initWithState:(NSString *)state {
    if ((self = [super init])) {
        _state = [state copy];
    }
    return self;
}

- (MementoSnapshot *)save {
    return [[MementoSnapshot alloc] initWithState:self.state];
}

- (void)restore:(MementoSnapshot *)snapshot {
    self.state = snapshot.state;
}
@end

BOOL verifyMementoCanonical(void) {
    MementoDocument *document = [[MementoDocument alloc] initWithState:@"draft"];
    MementoSnapshot *snapshot = [document save];

    document.state = @"published";
    BOOL changed = [document.state isEqualToString:@"published"];
    BOOL snapshotPreserved = [snapshot.state isEqualToString:@"draft"];

    [document restore:snapshot];
    BOOL restored = [document.state isEqualToString:@"draft"];

    document.state = @"archived";
    BOOL independent = [snapshot.state isEqualToString:@"draft"];

    return changed && snapshotPreserved && restored && independent;
}

#ifdef MEMENTO_STANDALONE
int main(void) {
    @autoreleasepool {
        if (!verifyMementoCanonical()) {
            return 1;
        }
        puts("Objective-C Memento: passed");
        return 0;
    }
}
#endif

File: src/Functional/Haskell/Memento.hs
module Memento
  ( Document (..),
    MementoSnapshot (..),
    save,
    restore,
    verifyMementoCanonical,
    main,
  )
where

data Document = Document
  { documentText :: String
  }
  deriving (Eq, Show)

newtype MementoSnapshot = MementoSnapshot String
  deriving (Eq, Show)

save :: Document -> MementoSnapshot
save = MementoSnapshot . documentText

restore :: Document -> MementoSnapshot -> Document
restore document (MementoSnapshot text) = document {documentText = text}

verifyMementoCanonical :: Bool
verifyMementoCanonical =
  let draft = Document "draft"
      snapshot = save draft
      published = draft {documentText = "published"}
      restored = restore published snapshot
   in documentText published == "published"
        && documentText restored == "draft"
        && snapshot == MementoSnapshot "draft"

main :: IO ()
main =
  if verifyMementoCanonical
    then putStrLn "Haskell Memento: passed"
    else error "Haskell Memento: failed"
